### PR TITLE
spec: propose an evolution of closeOnExit

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
# Summary of the Pull Request
This pull request introduces a specification for the next generation of `closeOnExit`.

## References

#1364 #1770

## PR Checklist
* [x] Specs #2563
* [x] CLA signed
* [ ] Tests added/passed/it is a spec
* [x] Requires documentation to be updated
* [x] I've discussed this with core contributors already

## Detailed Description of the Pull Request / Additional comments

Look at the spec.

## Validation Steps Performed

It is a spec.

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
TODO:

* [ ] create an issue
* [ ] specify a followup change to conhost code that will propagate its subprocess's exit code

**Comment by glen-84** [comment]
Shouldn't `!= 0` be `= 0`?

You want it to close when there is a successful (0) exit code? Or am I getting mixed up?

**Comment by mcpiroman** [comment]
I also find this confusing. For me the default should be that
success -> exit
failature -> stay opened.

But you could also support the opposite. Say someone has long running job in the background and if it fails he can immidiately see that terminal / tab closes (it's less important if it has just finished).

**Comment by DHowett** [comment]
@glen-84 correct: if `closeOnExit` is set to `graceful`, the tab will automatically close only if the connected application exits "gracefully" or successfully. The idea is that an application that runs and instantly terminates--or an application that stops accepting user input because it's died--probably put something useful on the screen for troubleshooting purposes, but an application that has exited successfully does not need any troubleshooting.

If the default switched the other way (close if `!= 0`, stay open if `0`) would be instantly frustrating, because a user would type `exit` in their shell of choice and we'd just sit there telling them that the shell exited. Like they requested. Back before we introduced `closeOnExit`, people complained so much about that! :grin:

**Comment by DHowett-MSFT** [comment]
@mcpiroman that is what is suggested here (default=`graceful`, close tab when exit code `== 0`)

**Comment by mcpiroman** [comment]
but there could be another option, `failature`: close when exit code `!= 0`

**Comment by glen-84** [comment]
@DHowett Since the spec text was not updated, I'm not sure if you realized that I was pointing out an issue on line `#22`. It should be `== 0`, and not `!= 0`. 🙂

**Comment by DHowett-MSFT** [comment]
@glen-84 I definitely misread your comment :smile:

File: doc/specs/#xxxx - closeOnExit evolution.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+## Inspiration
+
+Other terminal emulators like ConEmu have a similar feature.
+
+## Solution Design
+
+1. The existing `closeOnExit` profile key will become an enumerated string key with the following values/behaviours:
+   * `always` - a tab or pane hosting this profile will always be closed when the launched connection terminates.
+   * `graceful` - a tab or pane hosting this profile will only be closed **if** the launched connection closes with an exit code != 0.
```
Comment (DHowett-MSFT) [inline doc/specs/#xxxx - closeOnExit evolution.md]:
```suggestion
   * `graceful` - a tab or pane hosting this profile will be closed **iff** the launched connection closes with an exit code == 0.
```

File: doc/specs/#xxxx - closeOnExit evolution.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+## Inspiration
+
+Other terminal emulators like ConEmu have a similar feature.
+
+## Solution Design
+
+1. The existing `closeOnExit` profile key will become an enumerated string key with the following values/behaviours:
+   * `always` - a tab or pane hosting this profile will always be closed when the launched connection terminates.
+   * `graceful` - a tab or pane hosting this profile will be closed **iff** the launched connection closes with an exit code == 0.
```
Comment (glen-84) [inline doc/specs/#xxxx - closeOnExit evolution.md]: `iff`?

**Comment by DHowett-MSFT** [comment]
Shorthand for “if and only if”

**Comment by glen-84** [comment]
> Shorthand for “if and only if”

Oh ... I feel like I should have known that, even though it seems fairly uncommon in programming. 🙂

iff

**Comment by mu88** [comment]
Is there already an issue that will implement the specified *graceful* behavior? I'd love to see this feature and an issue to subscribe would be nice.

**Comment by DHowett-MSFT** [comment]
I am clearing the review status of all signed-off reviewers, and adding @mcpiroman.

@mcpiroman: This spec is inspired by your work in #2091, with some minor refinements. Thank you. :smile:

File: doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    * enum `TerminalConnection::ConnectionState`
+        * This enum attempts to encompass all potential connection states, even those which do not make sense for a local terminal.
+        * The wide variety of values will be useful to indicate state changes in a user interface.
+        * `NotConnected`: All new connections will start out in this state
+        * `Connecting`: The connection has been initated, but has not yet completed connecting.
+        * `Connected`: The connection is active.
+        * `Closing`: The connection is being closed (usually by request).
+        * `Closed`: The connection has been closed, either by request or from the remote end terminating successfully.
+        * `Failed`: The connection was unexpectedly terminated.
+            * A connection entering the `Failed` state will usually produce a message containing a reason.
```
Comment (DHowett-MSFT) [inline doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md]: potential note: a compliant connection should never go backwards in this state list.

File: doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    * event arguments `StateChangedEventArgs(ConnectionState)`
+    * typed event `StateChanged`, whose payload is a `StateChangedEventArgs`
+    * event `TerminalDisconnected` will be removed, as it is replaced by `StateChanged`
+3. As `TerminalApp` is responsible for producing connection instances, _it_ will subscribe to the new connection state event.
```
Comment (miniksa) [inline doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md]: Is the subscription forced then as a part of instantiation?

File: doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+## Inspiration
+
+Other terminal emulators like ConEmu have a similar feature.
+
+## Solution Design
+
+1. The existing `closeOnExit` profile key will become an enumerated string key with the following values/behaviours:
+   * `always` - a tab or pane hosting this profile will always be closed when the launched connection terminates.
+   * `graceful` - a tab or pane hosting this profile will be closed **iff** the launched connection closes with an exit code == 0.
```
Comment (carlos-zamora) [inline doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md]: I'm guessing an exit code of 0 means that it was expected to close?

So if an unexpected error/failure occurs, the pane will remain open?

Say I set the `commandline` setting to a faulty `.bat` file that decides to fail, does that mean that the pane will remain open with the details of why my `.bat` file failed?

File: doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+Windows Terminal will no longer immediately terminate on startup if the user's shell doesn't exist.
+
+### Compatibility
+
+There is an existing `closeOnExit` _boolean_ key that a user may have configured in profiles.json. The boolean values should map as follows:
+
+* `true` -> `graceful`
+* `false` -> `never`
+
+This will make for a clean transition to Windows Terminal's sane new defaults.
```
Comment (carlos-zamora) [inline doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md]: Will we ever consider deprecating support for the `bool` version of `closeOnExit`?

File: doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+## Potential Issues
+
+There will be no impact to Performance, Power or Efficiency.
+
+## Future considerations
+
+* Eventually, we may want to implement a feature like "only close on graceful exit if the shell was running for more than X seconds". This puts us in a better position to do that, as we can detect graceful and clumsy exits more readily.
+   * (potential suggestion: `{ "closeOnExit": "10s" }`
+* The enumerator values for transitioning connection sattes will be useful for connections that require internet access.
```
Comment (carlos-zamora) [inline doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md]:
```suggestion
* The enumerator values for transitioning connection states will be useful for connections that require internet access.
```

File: doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+This will make for a clean transition to Windows Terminal's sane new defaults.
+
+### Performance, Power, and Efficiency
+
+## Potential Issues
+
+There will be no impact to Performance, Power or Efficiency.
+
+## Future considerations
```
Comment (carlos-zamora) [inline doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md]: Just throwing some random ideas out there:
- **UI Ideas**:
  - make the tab flash for attention when the state changes to `Connected` (and maybe red when it becomes `Failed`)
  - While in the `Connecting` or `Closing` state, put a little spinning wheel in the tab to denote that it's working on it.

File: doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md (hunk 6 above)
Comment (zadjii-msft) [reply doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md]: I know there's another issue around here tracking deprecating old keys...

Comment (DHowett-MSFT) [reply doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md]: I honestly think true/false/never/graceful/always is fine to keep forever, but fair point.

File: doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md (hunk 5 above)
Comment (DHowett-MSFT) [reply doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md]: Correct.

File: doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    * event arguments `StateChangedEventArgs(ConnectionState)`
+    * typed event `StateChanged`, whose payload is a `StateChangedEventArgs`
+    * event `TerminalDisconnected` will be removed, as it is replaced by `StateChanged`
+3. As `TerminalApp` is responsible for producing connection instances, _it_ will subscribe to the new connection state event.
+
+The new default value for `closeOnExit` will be `graceful`.
+
+## UI/UX Design
+
+As above. An affordance is provided for the application consuming a connection to determine what state the connection is in and provide sufficient UI. Terminal will not currently make use of any of the new states other than `Closed` and `Failed` to enact UI change.
```
Comment (zadjii-msft) [inline doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md]: > Terminal will not currently make use of any of the new states other than `Closed` and `Failed` to enact UI change.

Do you want to call out in this spec what that UI would be like? As a user, when there's an ungraceful exit, what happens?

When it's set to `never`, and there's a graceful exit, what happens?

File: doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md (hunk 4 above)
Comment (DHowett-MSFT) [reply doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md]: It's not required, and I want to change the language here. I might pipe this up through TerminalControl so it mirrors its connection state and remove the TerminalControl's Closed event.

File: doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md (hunk 9 above)
Comment (DHowett-MSFT) [reply doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md]: I've filled out the UI/UX considerations section.

**Pull request merged**